### ISSUE TITLE
[Feature]Adding automatic model average parallelism support in TF

### ISSUE THREAD

**Issue by yangjunpro** [tensorflow/tensorflow#10449]
Model Average is a common paradigm for distributed DL training, also there are several papers regarding to it and its variants:

(https://arxiv.org/pdf/1410.7455v8.pdf)
(http://www.microsoft.com/en-us/research/wp-content/uploads/2016/08/0005880.pdf)

Do we have any plan to bring this support into TF? Actually we have already implemented model average support and benchmarked it on several in-house models, the speed up is good, about 3X(4 cards) to 9X(16 cards) convergence speed-up. Also we are working on making the model average mechanism as automatic as possible, so modeling guys could easily leverage this nice feature.

If TF community is fine with this feature/enhancement, we will be happy to merge our code into community.

Thanks.

**Comment by michaelisard** [comment]
@sherrym who is a good contact for figuring out how to work on this contribution?

**Comment by byronyi** [comment]
Interesting work! In the papers you posted it seems that's a new synchronization mechanism to reduce cross device communication cost.

As a side note, does your implementation change the core/distributed runtime or it is currently implemented above TF API (e.g. using Python)? Any significant differences from the synchronization methods mentioned [here](https://www.tensorflow.org/performance/performance_models#variable_distribution_and_gradient_aggregation)?

I'm just a little worried if that brings significant changes to the distributed runtime. The runtime is currently oblivious to the semantic of tensors it send/recv.

**Comment by yangjunpro** [comment]
@byronyi 
Currently, our implementation only involve Python-level change since we formulate this problem as a graph transformation problem so as to ensure our code change is orthogonal to low-level changes.

**Comment by byronyi** [comment]
I don't know about other members but as for myself I would love to see new synchronisation mechanism beyond the common ones we have right now (parameter server, distributed replicated using all-reduce, etc.). 

Looking forward to your patch.

**Comment by hellolovetiger** [comment]
@yangjunpro I recently implemented the distributed DL training method based on the paper you mentioned. Here is the details: #11071 .
But there are some problems with my implementation: the training loss decreases at the begining but will increases dramatically suddenly after the ps params do several updates.

**Comment by yangjunpro** [comment]
@hellolovetiger You could take a look at the PR in https://github.com/tensorflow/tensorflow/pull/11581

**Comment by hellolovetiger** [comment]
Cool, thanks, @yangjunpro .

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by michaelisard** [comment]
Closing this since the referenced PR is closed. Please reopen if there's something that should be done.
